File: apps/android/core/model/src/main/kotlin/work/kumarfamilynet/cinemarchive/core/model/RatingNormalization.kt
package work.kumarfamilynet.cinemarchive.core.model

import java.text.Collator
import kotlin.math.sqrt

enum class RatingBaseline { ALL, MEDIA }

/** One current title rating, never one observation per viewing or episode. */
data class RatingObservation(val titleId: String, val title: String, val type: MediaType, val rating: Double?)

data class RatingReference(
    val type: MediaType?,
    val label: String,
    val count: Int,
    val mean: Double?,
    val deviation: Double?,
    val percentiles: Map<Double, Double>,
)

data class NormalizedRating(
    val title: RatingObservation,
    val zScore: Double?,
    val percentile: Double,
    val baseline: RatingReference,
)

data class RatingNormalization(val groups: List<RatingReference>, val rows: List<NormalizedRating>)

/** Mirrors web ratingNormalization.ts: population SD, empirical midrank, and display-only scope. */
fun deriveRatingNormalization(
    titles: List<RatingObservation>,
    baseline: RatingBaseline = RatingBaseline.ALL,
    scope: String = "all",
): RatingNormalization {
    val rated = titles.filter { it.rating?.let { r -> r.isFinite() && r in 0.0..5.0 } == true }
    fun summarize(type: MediaType?, label: String): RatingReference {
        val ratings = rated.filter { type == null || it.type == type }.map { it.rating!! }
        val mean = ratings.takeIf { it.isNotEmpty() }?.average()
        val deviation = mean?.let { m -> sqrt(ratings.sumOf { (it - m) * (it - m) } / ratings.size) }
        var below = 0
        val percentiles = ratings.groupingBy { it }.eachCount().toSortedMap().mapValues { (_, tied) ->
            val rank = 100.0 * (below + tied / 2.0) / ratings.size
            below += tied
            rank
        }
        return RatingReference(type, label, ratings.size, mean, deviation, percentiles)
    }
    val groups = if (baseline == RatingBaseline.ALL) listOf(summarize(null, "All titles")) else listOf(
        summarize(MediaType.MOVIE, "Films"), summarize(MediaType.TV, "Series"),
    )
    val collator = Collator.getInstance()
    val rows = rated.filter { scope == "all" || it.type == if (scope == "movies") MediaType.MOVIE else MediaType.TV }
        .map { title ->
            val group = groups.first { it.type == null || it.type == title.type }
            val z = if (group.count >= 2 && (group.deviation ?: 0.0) > 1e-12) {
                (title.rating!! - group.mean!!) / group.deviation!!
            } else null
            NormalizedRating(title, z, group.percentiles.getValue(title.rating!!), group)
        }.sortedWith(
            compareByDescending<NormalizedRating> { it.zScore ?: Double.NEGATIVE_INFINITY }
                .thenByDescending { it.percentile }
                .thenComparator { a, b -> collator.compare(a.title.title, b.title.title) },
        )
    return RatingNormalization(groups, rows)
}

File: apps/web/src/store/ratingNormalization.ts
import type { Title } from './mockData'
import type { LedgerScope } from 'src/lib/ledgerPanels'

export type RatingBaseline = 'all' | 'media'
type RatedTitle = Title & { rating: number }

export function isRatedTitle(title: Title): title is RatedTitle {
  return typeof title.rating === 'number' && Number.isFinite(title.rating)
    && title.rating >= 0 && title.rating <= 5
}

function summarize(titles: RatedTitle[]) {
  const count = titles.length
  const mean = count ? titles.reduce((sum, title) => sum + title.rating, 0) / count : null
  // Population SD: describing this library, not estimating a wider population.
  const deviation = mean === null ? null : Math.sqrt(
    titles.reduce((sum, title) => sum + (title.rating - mean) ** 2, 0) / count,
  )
  const frequencies = new Map<number, number>()
  for (const title of titles) frequencies.set(title.rating, (frequencies.get(title.rating) ?? 0) + 1)
  const percentiles = new Map<number, number>()
  let below = 0
  for (const [rating, tied] of [...frequencies].sort(([a], [b]) => a - b)) {
    // Empirical midrank: ties share the middle of their occupied percentile range.
    percentiles.set(rating, 100 * (below + tied / 2) / count)
    below += tied
  }
  return { count, mean, deviation, percentiles }
}

/** One observation per title. Rewatches/episode logs do not receive extra weight.
 * Scope filters the displayed rows only; it never changes the chosen baseline.
 * Callers pass the currently visible owner's library, including sharing restrictions.
 */
export function deriveRatingNormalization(
  titles: Title[],
  baseline: RatingBaseline = 'all',
  scope: LedgerScope = 'all',
) {
  const rated = titles.filter(isRatedTitle)
  const groups = baseline === 'all'
    ? [{ key: 'all' as const, label: 'All titles', ...summarize(rated) }]
    : [
        { key: 'movie' as const, label: 'Films', ...summarize(rated.filter((t) => t.type === 'movie')) },
        { key: 'tv' as const, label: 'Series', ...summarize(rated.filter((t) => t.type === 'tv')) },
      ]
  const rows = rated
    .filter((t) => scope === 'all' || t.type === (scope === 'movies' ? 'movie' : 'tv'))
    .map((title) => {
      const group = groups.find((g) => g.key === 'all' || g.key === title.type)!
      const zScore = group.count >= 2 && group.deviation !== null && group.deviation > 1e-12
        ? (title.rating - group.mean!) / group.deviation
        : null
      return { title, zScore, percentile: group.percentiles.get(title.rating)!, baseline: group }
    })
    .sort((a, b) => (b.zScore ?? -Infinity) - (a.zScore ?? -Infinity)
      || b.percentile - a.percentile || a.title.title.localeCompare(b.title.title))
  return { groups, rows }
}
